### ISSUE TITLE
Support gdbstub targets

### ISSUE THREAD

**Issue by osandov** [osandov/drgn#172]
The [GDB Remote Serial Protocol](https://sourceware.org/gdb/onlinedocs/gdb/Remote-Protocol.html) is implemented by several debugging targets, including [kgdb for the Linux kernel](https://www.kernel.org/doc/html/latest/dev-tools/kgdb.html) and [QEMU](https://wiki.qemu.org/Features/gdbstub). We should support targets using this protocol, which will require implementing alternate backends for at least the following components of drgn:

* Debug info discovery.
* Memory reading.
* Threads.
* Thread events (see #171).

There are probably more things that would need gdbstub support, so this project is fairly open-ended.

**Comment by bgamari** [comment]
This would also likely unlock support for https://github.com/rr-debugger/rr, which I have found to be an invaluable tool when debugging user-space issues.

**Comment by brenns10** [comment]
One of the open questions for many of these options is going to be, how can we get the vmcoreinfo metadata for these targets. With #396, we will be able to let people manually set the vmcoreinfo prior to attaching, but it would be better if drgn could get that information automatically from the target.

Maybe a good place to get started is to add the ability to get vmcoreinfo from some of these backends where possible? EG QEMU's gdbstub could be extended to provide the cached vmcoreinfo if the guest was started with `-device vmcoreinfo`. I'm sure there must be some place in the GDB protocol to be able to ask for some custom piece of information...

It may be possible to get kgdb to provide a similar management command too?

**Comment by osandov** [comment]
Yeah, exactly, I was hoping that it'd be possible to add a custom GDB protocol message for this.

In the meantime, in addition to manually providing the `VMCOREINFO`, if KASLR is disabled and we're able to provide a vmlinux earlier in the program initialization process, we can get the vmcoreinfo note from symbols or DWARF. Terrible user experience, but probably unavoidable at the moment.

**Comment by t-8ch** [comment]
> I'm sure there must be some place in the GDB protocol to be able to ask for some custom piece of information...

See https://sourceware.org/gdb/current/onlinedocs/gdb.html/General-Query-Packets.html , https://github.com/torvalds/linux/blob/ffd294d346d1/kernel/debug/kdb/kdb_io.c#L420-L426 and `gdb_extend_query_table()`/`GdbCmdParseEntry` in QEMU.

For example a custom query `qlinux.vmcoreinfo` shared by all three backends should be doable.
(I have no direct experience with such custom queries, though)
